circuit/src/dynamic_sets/native.rs

```diff
@@ -145,109 +156,98 @@ impl<const NUM_NEIGHBOURS: usize, const NUM_ITERATIONS: usize, const INITIAL_SCO
 	}
 
 	/// Remove the member and its opinion
-	pub fn remove_member(&mut self, pk: PublicKey) {
+	pub fn remove_member(&mut self, pk: Fr) {
 		let pos = self.set.iter().position(|&(x, _)| x == pk);
 		// Make sure already in the set
 		assert!(pos.is_some());
 
 		let index = pos.unwrap();
-		self.set[index] = (PublicKey::default(), Fr::zero());
+		self.set[index] = (Fr::zero(), Fr::zero());
 
 		self.ops.remove(&pk);
 	}
 
 	/// Update the opinion of the member
-	pub fn update_op(&mut self, from: PublicKey, op: Opinion<NUM_NEIGHBOURS>) {
-		let pos_from = self.set.iter().position(|&(x, _)| x == from);
+	pub fn update_op(&mut self, from: ECDSAPublicKey, op: Vec<SignedAttestation>) -> Fr {
+		let from_pk = recover_ethereum_address_from_pk(from);
+		let pos_from = self.set.iter().position(|&(x, _)| x == from_pk);
 		assert!(pos_from.is_some());
 
-		self.ops.insert(from, op);
-	}
+		let mut scores = vec![Fr::zero(); NUM_NEIGHBOURS];
+		let mut hashes = Vec::new();
+		for (i, att) in op.iter().enumerate() {
+			let is_default_pubkey = self.set[i].0 == Fr::zero();
+
+			if is_default_pubkey {
+				scores[i] = Fr::default();
+				hashes.push(AttestationFr::default().hash());
```

Comment:
`AttestationFr::default().hash()` can be defined before the loop, and reused.

The rest looks good